### PR TITLE
RavenDB-5499 & RavenDB-5609

### PR DESCRIPTION
**Pull request by redknightlois**

### CODE REVIEW

File: src/Voron/Impl/Paging/AbstractPager.cs [hunk 1]
```diff
@@ -185,6 +189,17 @@ public PagerState EnsureContinuous(long requestedPageNumber, int numberOfPages)
             return AllocateMorePages(allocationSize);
         }
 
+        [Conditional("VALIDATE")]
```
Comment (ayende) [inline src/Voron/Impl/Paging/AbstractPager.cs:192]: Does the conditional property applies to derived methods?

File: src/Voron/Platform/Win32/Win32MemoryMapPager.cs [hunk 2]
```diff
@@ -436,5 +454,49 @@ public override void MaybePrefetchMemory(List<TreePage> sortedPages)
...
+                Win32NativeMethods.MemoryProtection oldProtection;
+                bool status = Win32NativeMethods.VirtualProtect(start, new UIntPtr(size), Win32NativeMethods.MemoryProtection.READONLY, out oldProtection);
+                if (!status)
+                {
+                    int vProtectError = Marshal.GetLastWin32Error();
+
+                    Win32NativeMethods.MEMORY_BASIC_INFORMATION memoryInfo2 = new Win32NativeMethods.MEMORY_BASIC_INFORMATION();
+                    int vQuerySecondOutput = Win32NativeMethods.VirtualQuery(start, &memoryInfo2, new UIntPtr(size));
+                    int vQuerySecondError = Marshal.GetLastWin32Error();
+                    Debugger.Break();
```
Comment (ayende) [inline src/Voron/Platform/Win32/Win32MemoryMapPager.cs:475]: I assume that this is only ever called IFF the VALIDATION is setup, right?

File: src/Voron/Impl/Paging/AbstractPager.cs (hunk 1 above)
Comment (jbayardo) [reply src/Voron/Impl/Paging/AbstractPager.cs:192]: Yes. 17.4.2.1 of the C# spec says: "The conditional method must not be marked with the override modifier. A conditional method may be marked with the virtual modifier, however. Overrides of such a method are implicitly conditional, and must not be explicitly marked with a Conditional attribute."

File: src/Voron/Platform/Win32/Win32MemoryMapPager.cs (hunk 2 above)
Comment (jbayardo) [reply src/Voron/Platform/Win32/Win32MemoryMapPager.cs:475]: See above, but yes.

Comment (redknightlois) [reply src/Voron/Platform/Win32/Win32MemoryMapPager.cs:475]: Debugger.Break while an interesting thing to happen, will have impact on CI server if we have such an error.  I would remove it.

**Comment by ayende** [comment]
So probably better to throw

*Hibernating Rhinos Ltd  *

Oren Eini\* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Pull request merged**